### PR TITLE
tbb: 2020.3 -> 2021.7.0

### PR DESCRIPTION
**Pull request by erictapen**
###### Description of changes

It builds for `x86_64-linux`, but there is still an error for Musl:

```
In file included from /tmp/nix-build-tbb-2021.7.0.drv-0/source/test/common/test.h:32,
                 from /tmp/nix-build-tbb-2021.7.0.drv-0/source/test/tbbmalloc/test_malloc_overload.cpp:45:
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/tbbmalloc/test_malloc_overload.cpp: In function ‘void DOCTEST_ANON_FUNC_41()’:
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/tbbmalloc/test_malloc_overload.cpp:400:13: error: ‘mallopt’ was not declared in this scope; did you mean ‘malloc’?
  400 |     REQUIRE(mallopt(0, 0)); // add dummy mallopt call for coverage
      |             ^~~~~~~
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/common/doctest.h:1997:9: note: in definition of macro ‘DOCTEST_WRAP_IN_TRY’
 1997 |         x;                                                                                         \
      |         ^
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/common/doctest.h:2249:9: note: in expansion of macro ‘DOCTEST_ASSERT_IMPLEMENT_2’
 2249 |         DOCTEST_ASSERT_IMPLEMENT_2(assert_type, __VA_ARGS__);                                      \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/common/doctest.h:2268:30: note: in expansion of macro ‘DOCTEST_ASSERT_IMPLEMENT_1’
 2268 | #define DOCTEST_REQUIRE(...) DOCTEST_ASSERT_IMPLEMENT_1(DT_REQUIRE, __VA_ARGS__)
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/common/doctest.h:2743:22: note: in expansion of macro ‘DOCTEST_REQUIRE’
 2743 | #define REQUIRE(...) DOCTEST_REQUIRE(__VA_ARGS__)
      |                      ^~~~~~~~~~~~~~~
/tmp/nix-build-tbb-2021.7.0.drv-0/source/test/tbbmalloc/test_malloc_overload.cpp:400:5: note: in expansion of macro ‘REQUIRE’
  400 |     REQUIRE(mallopt(0, 0)); // add dummy mallopt call for coverage
      |     ^~~~~~~
compilation terminated due to -Wfatal-errors.
make[2]: *** [test/CMakeFiles/test_malloc_overload.dir/build.make:76: test/CMakeFiles/test_malloc_overload.dir/tbbmalloc/test_malloc_overload.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:4110: test/CMakeFiles/test_malloc_overload.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Also I didn't test this on any other platform.

###### Changelogs:

- [2021.1.1](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)
- [2021.2.0](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)
- [2021.3.0](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)
- [2021.4.0](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)
- [2021.5.0](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)
- [2021.6.0](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)
- [2021.7.0](https://github.com/oneapi-src/oneTBB/blob/v2021.7.0/RELEASE_NOTES.md)

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
See also https://github.com/NixOS/nixpkgs/pull/109113, has been a lot of breakage in Nixpkgs and also https://github.com/lopsided98/nix-ros-overlay (e.g. https://github.com/osrf/gazebo/issues/2867).

**Comment by chuangzhu** [comment]
> there has been a lot of breakage in Nixpkgs

How about packaging multiple versions like `tbb_2020` and `tbb_2021`?

File: pkgs/development/libraries/tbb/default.nix [hunk 1]
```diff
@@ -1,88 +1,28 @@
...
-    (fetchurl {
-      name = "aarch64-darwin.patch";
-      url = "https://github.com/oneapi-src/oneTBB/pull/258/commits/86f6dcdc17a8f5ef2382faaef860cfa5243984fe.patch";
-      sha256 = "sha256-JXqrFPCb3q1vfxk752tQu7HhApCB4YH2LoVnGRwmspk=";
-    })
-  ];
-
-  nativeBuildInputs = lib.optionals stdenv.isDarwin [
+  nativeBuildInputs = [ cmake pkg-config ] ++ (lib.optionals stdenv.isDarwin [
     fixDarwinDylibNames
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/tbb/default.nix:21]:
```suggestion
  nativeBuildInputs = [ cmake pkg-config ]
    ++ lib.optional stdenv.isDarwin fixDarwinDylibNames;
```

File: pkgs/development/libraries/tbb/default.nix [hunk 2]
```diff
@@ -1,88 +1,28 @@
...
-  ];
-
-  makeFlags = lib.optionals stdenv.cc.isClang [
-    "compiler=clang"
-  ];
+  ]);
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/tbb/default.nix:22]:
```suggestion
```

File: pkgs/development/libraries/tbb/default.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/development/libraries/tbb/default.nix:21]: Always using `optionals` and multi-line dependencies makes it less error-prone and more `git diff` and `blame` friendly.

**Comment by NickCao** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/217585

**Pull request closed without merging**